Add administration utilities to AVL codebase

pont-us opened the pull request:
Closes #7 .

This PR adds the a new module `avl._admin` containing the following classes:

 - `AwsResourceCreator`, which creates the persistent AWS resources required by the AVL (previously done by hand in the AWS web console).
 - `BucketAccessUserCreator`, intended to be used by the Jupyter Hub process to create (possibly temporary) IAM users for access to the AVL user bucket (expanded and improved from the current code in https://github.com/agriculture-vlab/avl-jh-hub-image ).

pont-us committed:
Add AVL _admin module, resource creator, and tests

Add a new module "_admin" for code related to administration and
management of AVL infrastructure. At present it contains the
class AvlResourceCreator, which creates AWS resources for the AVL
deployment.

pont-us committed:
Rename AvlResourceCreator to AwsResourceCreator

pont-us committed:
Update environment.yml

 - Add boto3 and moto dependencies
 - Bump minimum xcube version to 0.10.1

pont-us committed:
Add BucketAccessUserCreator to _admin module

This class is designed to be used by the Jupyter Hub process when
spawning a new user environment, to create an IAM user for access to
the user bucket. It's based on the create_user script currently
residing in the AVL user hub custom image repository.

pont-us committed:
Add some to-do comments in the _admin module

pont-us committed:
Expand and improve admin code and tests

 - Add more parameters to administration methods to avoid hardcoding
   things like resource name prefixes.

 - When creating bucket access users and access keys, delete old access
   keys first to avoid hitting AWS limits.

 - Test behaviour more thoroughly using moto mocks.

pont-us committed:
Add docstrings for resource creation code

pont-us committed:
Minor improvements to administration code & tests

 - Remove print statements or turning them into logging calls.

 - Add some more logging calls to the user creation process.

 - Remove some redundant code in a unit test.

pont-us committed:
AwsResourceCreator: update policy for user manager

The user manager IAM policy gives it permission to operate only on
user IDs under the AVL path, but this path was incorrectly specified
in the policy. Now fixed.

AliceBalfanz commented on avl/_admin.py:
```diff
@@ -0,0 +1,377 @@
+
+    This class is intended to be used by the Jupyter Hub process when
+    spawning a new user environment. It creates (if not already present)
+    a temporary IAM user with permissions to read and write to the AVL
```
```suggestion
    an  IAM user with permissions to read and write to the AVL
```
If I remember correctly, we decided, that the IAM users are not temporary, but created once and then kept?

pont-us replied:
I think we left it open -- inactive IAM users can in principle be culled at any time without any noticeable effect, because they are recreated whenever needed, but at present the culling isn't yet implemented because there aren't enough users to make it necessary. But you're right, the term "temporary" is probably confusing in this context because IAM users aren't temporary in the same sense as roles, which have a predefined, limited lifespan. I'll update the documentation to make this clearer.

pont-us replied:
I've replaced "temporary user" with "bucket access user" throughout, and added more clarification about how the IAM users are created and (possibly) deleted.

AliceBalfanz commented on avl/_admin.py:
```diff
@@ -0,0 +1,377 @@
+        :param user_name: the AVL user name, which will be used as one
+               component of the IAM user name
+        :param client_id: the client ID of the IAM user that will be used to
+               create the temporary user. It is expected that this will be
```
again, sure it should be temporary here?

pont-us replied:
Probably not, will fix -- see longer comment above.

AliceBalfanz commented on avl/_admin.py:
```diff
@@ -0,0 +1,377 @@
+    def create_user(self) -> Tuple[str, str]:
+        """Create a temporary bucket access user (if required) and access key
+
+        If the user already exists, a new access key and secret are created.
```
Does this mean, that the AVL user get a new pair of access credentials each session?

pont-us replied:
Yes! But they probably won't notice, since the credentials are set in the standard AWS environment variables so they're used automatically by s3fs without having to be passed as arguments anywhere.

pont-us replied:
I've made it more explicit in the docstring that new credentials are returned on every call. I've also renamed the method to `ensure_user_and_create_key`.

AliceBalfanz commented on environment.yml:
```diff
@@ -2,13 +2,15 @@ name: agriculture-vlab
   # Testing
   - pytest >=4.4
   - pytest-cov >=2.6
+  - moto >=3.0.5
```
```suggestion
  - moto >=3.0.5
  
```

AliceBalfanz reviewed (approved):
I left some comments  concerning "temporary". I have not marked each occurrence, but I would like clarification before the PR is merged :) 

Thanks a lot!

pont-us committed:
Improve docstrings and rename a method

 - Expand and improve docstrings. In particular, use the term
   "bucket access user" instead of "temporary user", since these IAM
   users don't *necessarily* have to be temporary, and it's largely
   irrelevant to the BucketAccessUserCreator class if they are or not.

 - Rename the method BucketAccessUserCreator.create_user to the more
   accurate ensure_user_and_create_key.
